### PR TITLE
[Scheduler] Add HRRN schedule policy to significantly reduce TTFT

### PR DESCRIPTION
**Pull request by SovietPower**
## Motivation

HRRN (Highest Response Ratio Next) is a classic scheduling algorithm. Like SJF it prioritizes short-service requests to reduce queueing latency, but it additionally weights each request by how long it has waited, so no request can be starved by an unbounded stream of shorter ones. This makes HRRN theoretically optimal for mean response time while avoiding the tail-latency pathology of pure SJF variants. Prefill is a good fit because per-request work can be approximately estimated by the uncached input length, which makes HRRN feasible.

Two internal runs against real production traffic show that HRRN significantly outperforms the existing `fcfs` / `lpm` / `dfs-weight` policies on avg, p95, and throughput, while slightly reducing p99 (i.e. avoiding starvation).

| conc=100, TTFT (s) | avg  |         | p50   |         | p90   |         | p99   |         | input TPS |        |
| ------------------ | ---- | ------- | ----- | ------- | ----- | ------- | ----- | ------- | --------- | ------ |
| FCFS (baseline)    | 24.2 |         | 13.43 |         | 50.92 |         | 60.5  |         | 119023    |        |
| LPM                | 16.4 | -32.23% | 5.6   | -58.30% | 32.79 | -35.60% | 213.3 | +252.56% | 148458    | +24.73% |
| DFS-weight         | 19.0 | -21.49% | 6.98  | -48.03% | 45.34 | -10.96% | 164.1 | +171.24% | 141790    | +19.13% |
| HRRN               | 7.48 | -69.09% | 4.5   | -66.49% | 9.64  | -81.07% | 55.48 | -8.30%  | 166135    | +39.58% |

> Test environment: sglang v0.5.15.post1, GLM-5.2 nvfp4, 1P1D on 2×8×B200. Prefill uses `attn-cp=8`, `hicache`, `chunk-size=65536`. Decode uses `dp-attention=8`. Traffic is real production (mixed agent workloads with up to 1M-token inputs).

Uncached input-length distribution across the test corpus (min, p10, p20, …, p90, p95, p99, max):

```
1  71  259  476  1047  6440  22058  36544  57981  93373  151406  257093  479058
```

Mean is 33927. The distribution has huge dispersion and a long tail — the regime where HRRN's response-ratio formulation clearly beats FCFS / LPM / DFS-weight.

## Modifications

Two equivalent forms of HRRN were considered:

1. **Classic form**: `key = (wait_time + prefill_time) / prefill_time = 1 + wait_time / prefill_time`.
   - `wait_time` depends on `now` and `arrival_time`. Across TP ranks / schedulers these may diverge, so keys differ and sort results differ — would require rank-0-sorts-and-broadcasts to keep ranks consistent.
   - `prefill_time` depends on an estimated TPS, which needs a user-set parameter or online adjustment.

2. **Token-based form**: `key = 1 + waited_tokens / uncached_tokens`.
   - `waited_tokens` = number of prefill tokens the scheduler has processed since this request entered `waiting_queue`, proportional to wait time.
   - `wait_time = waited_tokens / TPS`, `prefill_time = uncached_tokens / TPS`, so this form is mathematically equivalent to (1).
   - No wall-clock dependency → sort output is identical across ranks by construction.
   - No TPS parameter → more robust to throughput fluctuation, and arguably more accurate than the classic form.
   - Simpler, but requires tracking `waited_tokens` per req.

Chose form (2). The scheduler keeps a running counter of prefill tokens processed so far; each req snapshots this counter at its arrival into `waiting_queue`; `waited_tokens = current_counter - arrival_snapshot`. `rid` is used as a tie-breaker so equal-ratio reqs sort deterministically and identically across ranks.

A note on the work estimator: using `uncached_tokens` alone as a proxy for per-request prefill cost is effective for sparse-attention models such as GLM-5, where per-token prefill cost is roughly context-length-independent. For other models (e.g. dense attention whose per-token cost grows with context length), a mixed estimator such as `a * cached_tokens + uncached_tokens` may be more accurate; the current implementation does not expose this yet, but the token-based formulation makes such an extension straightforward.

Enable via `--schedule-policy hrrn`.

### Notes

Variants such as `waited^γ / uncached` or `(waited + uncached) / uncached^β` may further tune the tradeoff between SUF strength and aging speed. Limited testing so far hasn't found a parameter setting that clearly beats the plain HRRN formula, so this PR ships the plain form only.

## Accuracy Tests

HRRN only affects the order in which prefill requests are picked; it does not change model forward or KV/logits computation, so per-request model outputs are unchanged. Verified on GLM-5.2-nvfp4 with the full GPQA-Diamond set (`sglang.test.run_eval --eval-name gpqa --num-examples 198 --repeats 3 --max-tokens 128000 --top-p 0.95 --temperature 1.0 --thinking-mode glm-4.5`):

| policy | GPQA-Diamond score |
| ------ | ------------------ |
| FCFS   | 0.906              |
| HRRN   | 0.9175             |

Difference is within sampling noise on 198 questions.

## Speed Tests and Profiling

The table in the Motivation section above is a summary. Full results of one internal-traffic benchmark run: https://my.feishu.cn/sheets/EZgXsbxx2hf4TDttt1sc2uYfnCh?from=from_copylink&sheet=be4e45

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #34121120891](https://github.com/sgl-project/sglang/actions/runs/34121120891)
Latest PR Test (Extra): :white_check_mark: [Run #34121120685](https://github.com/sgl-project/sglang/actions/runs/34121120685)
Latest PR Test (AMD ROCm 7.2): :x: [Run #34121120727](https://github.com/sgl-project/sglang/actions/runs/34121120727)

### CODE REVIEW

**Comment by SovietPower** [comment]
@hnyls2002 @merrymercy Could you please take a look or run-ci when convenient? This PR is a small, self-contained and effective addition to schedule_policy.py without changes to existing policies, and it's opt-in via --schedule-policy hrrn. We've been running this in production and hope to see it merged. Thank you!

File: python/sglang/srt/managers/scheduler.py [hunk 1]
```diff
@@ -3732,6 +3738,11 @@ def run_batch(
         batch.after_idle_gap = self._sched_idled
         self._sched_idled = False
 
+        # Accumulate the prefill-token counter used by the HRRN scheduling policy. Decode / prebuilt batches contribute 0.
+        _ent = getattr(batch, "extend_num_tokens", None) or 0
```
Comment (hnyls2002) [inline python/sglang/srt/managers/scheduler.py]: Can you avoid using `getattr`?

File: python/sglang/srt/server_args.py [hunk 2]
```diff
@@ -882,6 +882,7 @@ class ServerArgs:
                 "lof",
                 "priority",
                 "routing-key",
+                "hrrn",
```
Comment (BBuf) [inline python/sglang/srt/server_args.py:756]: It looks like `arrival_processed_tokens` is not set in unified mode. Should we either add it in `_add_request_to_queue`, or restrict HRRN to disaggregated prefill only?

File: python/sglang/srt/managers/schedule_policy.py [hunk 3]
```diff
@@ -288,7 +296,14 @@ def calc_priority(
...
 
     def _determine_active_policy(self, waiting_queue: List[Req]) -> Policy:
-        if self.policy == CacheAwarePolicy.LPM and len(waiting_queue) > 128:
+        if (
+            self.policy
+            in (
+                CacheAwarePolicy.LPM,
+                CacheAwarePolicy.HRRN,
+            )
+            and len(waiting_queue) > 128
```
Comment (BBuf) [inline python/sglang/srt/managers/schedule_policy.py:311]: Is it intended to fall back to FCFS when the waiting queue exceeds 128? This may disable HRRN under heavy load.

File: test/manual/test_schedule_policy.py [hunk 4]
```diff
@@ -326,6 +326,76 @@ def test_calc_priority_routing_key_empty_running_batch(self):
         self.assertEqual(waiting_queue[1].rid, "w2")
         self.assertEqual(waiting_queue[2].rid, "w3")
 
+    def test_calc_priority_hrrn(self):
```
Comment (BBuf) [inline test/manual/test_schedule_policy.py]: Could we move these tests to `test/registered` so they can run in CI?

File: python/sglang/srt/managers/schedule_policy.py (hunk 3 above)
Comment (SovietPower) [reply python/sglang/srt/managers/schedule_policy.py:311]: Yes, this is intentional — HRRN needs the same prefix-match pass as LPM to compute uncached lengths, so on very large queues the O(N·M) match may become expensive. We follow LPM's precedent here. Happy to lift the threshold or remove HRRN from this if needed.

File: python/sglang/srt/managers/scheduler.py (hunk 1 above)
Comment (SovietPower) [reply python/sglang/srt/managers/scheduler.py]: Updated in the new commit 31ac9c6b. Also dropped the two other getattr uses in the HRRN path — the fields are now always present on Req/ScheduleBatch

File: python/sglang/srt/server_args.py (hunk 2 above)
Comment (SovietPower) [reply python/sglang/srt/server_args.py:756]: Fixed in the new commit 31ac9c6b — snapshot is now also taken in `_add_request_to_queue` for the unified path.

File: test/manual/test_schedule_policy.py (hunk 4 above)
Comment (SovietPower) [reply test/manual/test_schedule_policy.py]: Moved the HRRN tests to `test/registered/unit/managers/test_schedule_policy.py` with `register_cpu_ci` in the latest commit. Left the other policy tests in `test/manual/` since they're out of scope here, and I also noticed #33909 is doing the broader migration for those.

**Comment by BBuf** [comment]
Please fix lint

**Review by hnyls2002 — APPROVED** [review]

**Review by BBuf — APPROVED** [review]
Thanks for your contribution.
